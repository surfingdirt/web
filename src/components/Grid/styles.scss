.col4 {
  .div {
    width: 25%;
  }
}
.fixedCol3,
.col3 {
  .div {
    width: 33.333%;
  }
}

.col2 {
  .div {
    width: 50%;
  }
}

%displayGrid {
  display: grid;
  grid-gap: 20px;
  margin: 20px 0;
}

@supports (display: grid) {
  .col4 {
    @extend %displayGrid;
    grid-template-columns: repeat(4, 1fr);
  }
  .fixedCol3,
  .col3 {
    @extend %displayGrid;
    grid-template-columns: repeat(3, 1fr);
  }
  .col2 {
    @extend %displayGrid;
    grid-template-columns: repeat(2, 1fr);
  }
  @media screen and (max-width: 750px){
    .col4,
    .col3 {
      grid-template-columns: repeat(1, 1fr);
    }
  }
}
