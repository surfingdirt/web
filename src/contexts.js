import Cookies from 'js-cookie';
import React from 'react';
import Login from '~/Login';
import routes from '~/routes';

const { HOME } = routes;
const LoginCookie = Login.COOKIE_NAME;
const signupDestinationKey = 'signUpDestination';

class AppContextValueObject {
  constructor(values) {
    this.values = {
      SSR: false,
      accountKey: '',
      availableLanguages: [],
      baseUrl: '',
      dir: 'ltr',
      facebookAppId: '',
      googleClientId: '',
      twitterUsername: '',
      graphql: '',
      language: '',
      login: {
        data: {
          accessToken: null,
          me: {
            id: null,
            username: '',
            email: '',
            firstname: '',
            lastname: '',
            gender: '',
            birthYear: '',
            country: '',
            phone: '',
            imageUrl: '',
            status: '',
            favoritesTeam: [],
          },
        },
      },
      nowValue: null,
      screenWidth: undefined,
      sportPlayer: { iframe: '', src: '' },
      translations: {},
    };

    Object.keys(values).forEach((key) => {
      if (this.values.hasOwnProperty(key)) {
        this.values[key] = values[key];
      } else {
        throw new Error(`Trying to assign unknown property: ${key}`);
      }
    });

    this.values.login.onSuccess = ({ accessToken, expires }) => {
      Cookies.set(LoginCookie, accessToken, { expires });

      const destination = this.values.login.getOrigin() || HOME;
      this.values.login.clearOrigin();
      window.location = destination;
    };

    this.values.login.onFailure = () => {
      this.values.login.data = {
        id: null,
        accessToken: null,
        expires: null,
        tokenType: null,
      };
    };

    this.values.login.logout = () => {
      Cookies.remove(LoginCookie);
      this.resetUser();
    };

    this.values.login.saveOrigin = () => {
      if (!global.window) {
        // Don't worry about that on the server.
        return;
      }
      sessionStorage.setItem(signupDestinationKey, window.location.pathname);
    };

    this.values.login.clearOrigin = () => {
      sessionStorage.removeItem(signupDestinationKey);
    };

    this.values.login.getOrigin = () => {
      if (!global.window) {
        // Don't worry about that on the server.
        return null;
      }
      return sessionStorage.getItem(signupDestinationKey);
    };
  }

  setAccessToken(accessToken) {
    this.values.login.data.accessToken = accessToken;
  }

  setUser({ id, username, email, bookmarks, likes, watches }) {
    this.values.login.data.me = { id, username, email, bookmarks, likes, watches };
  }

  resetUser() {
    this.setUser({
      id: null,
      username: '',
      email: '',
    });
  }

  getValues() {
    return this.values;
  }
}

const AppContext = React.createContext(AppContextValueObject);

export default {
  AppContext,
  AppContextValueObject,
};
