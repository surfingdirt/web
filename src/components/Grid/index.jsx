import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import styles from './styles.scss';

export default class Grid extends PureComponent {
  static propTypes = {
    ariaLabel: PropTypes.string,
    type: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    // Trying to say "renderer is a component" here, but there is no good way to do that:
    renderer: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
  };

  static defaultProps = {
    ariaLabel: '',
  };

  constructor(props) {
    super(props);

    this.renderItem = this.renderItem.bind(this);
  }

  renderItem(item) {
    const { renderer: Renderer } = this.props;
    return <Renderer key={item.id} item={item} />;
  }

  render() {
    const { ariaLabel, type, items } = this.props;

    switch (type) {
      case 'fixed-col3':
        return <div className={styles.fixedCol3}>{items.map(this.renderItem)}</div>;
      case 'col2':
        return (
          <div aria-label={ariaLabel} className={styles.col2}>
            {items.map(this.renderItem)}
          </div>
        );

      case 'col3':
        return (
          <div aria-label={ariaLabel} className={styles.col3}>
            {items.length === 4 || items.length === 5
              ? items.map((item, index) => index < 3 && this.renderItem(item))
              : items.map((item, index) => index <= 5 && this.renderItem(item))}
          </div>
        );
      case 'col4':
        return (
          <div aria-label={ariaLabel} className={styles.col4}>
            {items.map((item) => this.renderItem(item))}
          </div>
        );
      default:
        throw new Error(`Unsupported type grid type '${type}'`);
    }
  }
}

export const buildPlaceholderItemList = (count) => {
  return [...Array(count)].map((_, i) => ({ id: i }));
};
