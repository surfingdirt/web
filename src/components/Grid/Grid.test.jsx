import React from 'react';
import contexts from '~/contexts';

import Grid from './index';

const { AppContext } = contexts;

const items = [
  {
    id: '8dda478e-342b-44f1-b6d8-3a99278efbf0',
    name: 'ByOAo6wtG',
    description: null,
    url:'toto',
    poster: 'tata',
  },
  {
    id: '9bca784d-212a-32f9-d8b6-2b77648deae8',
    name: 'H1VXq1bX7',
    description: null,
    url: null,
    poster: 'titi',
  },
];

const contextValue = { availableLanguages: ['en'], language: 'en' };

const Card = ({ item }) => {
  return (
    <div>
      <img style={{ width: '100%' }} src={item.poster} alt={`item #${item.id}`} />
    </div>
  );
};

describe('Grid', () => {
  it('type "col3" should render a grid with 3 items', () => {
    const wrapper = mount(
      <AppContext.Provider value={contextValue}>
        <Grid type="col3" items={items} resourceType="video" renderer={Card} />
      </AppContext.Provider>,
    );
    expect(wrapper).toMatchSnapshot();

    const grid = wrapper.find('div.col3');
    expect(grid.children()).toHaveLength(3);
  });

  it('type "col4" should render a grid with 4 items', () => {
    const wrapper = mount(
      <AppContext.Provider value={contextValue}>
        <Grid type="col4" items={items} resourceType="video" renderer={Card} />
      </AppContext.Provider>,
    );
    expect(wrapper).toMatchSnapshot();

    const grid = wrapper.find('div.col4');
    expect(grid.children()).toHaveLength(4);
  });
});
