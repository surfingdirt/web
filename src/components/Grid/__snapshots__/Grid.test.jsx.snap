// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid type "col3" should render a grid with 3 items 1`] = `
<div
  aria-label=""
  className="col3"
>
  <div>
    <img
      alt="item #8dda478e-342b-44f1-b6d8-3a99278efbf0"
      src="https://assets.onrewind.tv/events-posters/7e418a39-6b39-4764-a27b-2daac01acd7b.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <img
      alt="item #9bca784d-212a-32f9-d8b6-2b77648deae8"
      src="https://assets.onrewind.tv/events-posters/219709c1-93d3-4bed-a681-369089dce12c_liveOff.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <img
      alt="item #38fd3e65-9f13-4135-acf0-1e274c83561e"
      src="http://s1.dmcdn.net/qSL4X.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;

exports[`Grid type "col4" should render a grid with 4 items 1`] = `
<div
  aria-label=""
  className="col4"
>
  <div>
    <img
      alt="item #8dda478e-342b-44f1-b6d8-3a99278efbf0"
      src="https://assets.onrewind.tv/events-posters/7e418a39-6b39-4764-a27b-2daac01acd7b.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <img
      alt="item #9bca784d-212a-32f9-d8b6-2b77648deae8"
      src="https://assets.onrewind.tv/events-posters/219709c1-93d3-4bed-a681-369089dce12c_liveOff.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <img
      alt="item #38fd3e65-9f13-4135-acf0-1e274c83561e"
      src="http://s1.dmcdn.net/qSL4X.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
  <div>
    <img
      alt="item #38fd3e65-9f13-4135-acf0-1e274c83561a"
      src="http://s1.dmcdn.net/qSL4X.jpg"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
